docs/development/design/build-images.rst

```diff
@@ -0,0 +1,274 @@
+we discarded ``stable``, old ``latest`` became the new ``stable`` and old ``testing`` became the new ``latest``.
+This produced issues to people pinning their images to any of these names because after this change,
+*we changed all the images for all the users* and many build issues arrised!
+
+
+Goals
+-----
+
+* release completely new Docker images without forcing users to change their pinned image
+* allow users to stick with an image "forever" (~years)
+* use a ``base`` image with the dependencies that don't change frequently (OS and base requirements)
+* ``base`` image naming is tied to the OS version (e.g. Ubuntu LTS)
+* allow us to update a Python version without affecting the ``base`` image
+* reduce size on builder VM disks by sharing Docker image layers
+* allow users to specify extra dependencies (apt packages, node, rust, etc)
+* allow use "limited" custom images for users by sharing most layers
+* automatically build & push *all* images on commit
+* deprecate ``stable``, ``latest`` and ``testing``
+
+
+New build image structure
+-------------------------
+
+.. Taken from https://github.com/readthedocs/readthedocs-docker-images/blob/master/Dockerfile
+
+* ``ubuntu20-base``
+
+  * labels
+  * environment variables
+  * system dependencies
+  * install requirements
+  * LaTeX dependencies (for PDF generation)
+  * other languages version managers (``pyenv``, ``nodenv``, etc)
+  * UID and GID
+
+The following images all are based on ``ubuntu20-base``:
+
+* ``ubuntu20-py*``
+
+  * Python version installed via ``pyenv``
+  * default Python packages (pinned versions)
+    * pip
+    * setuptools
+    * virtualenv
+  * labels
+
+* ``ubuntu20-conda*``
+
+  * same as ``-py*`` versions
+  * ``mamba`` executable
+
+Note that all these images only need to run ``pyenv install ${PYTHON_VERSION}``
+to install a specific Python/Conda version.
+
+.. Build all these images with Docker
+
+   docker build -t readthedocs/build:ubuntu20-base -f Dockerfile.base .
+   docker build -t readthedocs/build:ubuntu20-py39 -f Dockerfile.py39 .
+   docker build -t readthedocs/build:ubuntu20-conda47 -f Dockerfile.conda47 .
+
+   Check the shared space between images
+   docker system df --verbose | grep -E 'SHARED SIZE|readthedocs'
+
+   Initial Dockerfile.* as example for this are pushed in this PR
+   https://github.com/readthedocs/readthedocs-docker-images/pull/166
+
+
+Specifying extra users' dependencies
+------------------------------------
+
+Different users may have different requirements. We were already requested to install
+``swig``, ``imagemagick``, ``libmysqlclient-dev``, ``lmod``, ``rust``, ``poppler-utils``, etc.
+
+People with specific dependencies will be able to install them as APT packages or as extras
+using ``.readthedocs.yaml`` config file. Example:
+
+.. code:: yaml
+
+   build:
+     image: ubuntu20-py39
+     apt:
+       - swig
+       - imagemagick
+     extras:
+       - node==14.16
+       - rust==1.46.0
```

Comment:
also, maybe we just need to find a way to allow to the current user to use apt, so these can be done with any custom command (people may want to use a custom package from a repository that isn't in ubuntu)

Reply:
ansible does that easy with package modules, but you can restrict the package managers you want eg. https://github.com/staticdev/linux-developer-playbook/blob/main/default.config.yml#L51

Reply:
I know the plan is to maybe support nodeenv and rustenv. Maybe we consider not supporting those as a special case and we just have nodeenv/rustenv installed and they can run an arbitrary command to download and install the right version?

Reply:
I decided to treat them as special thinking that in the future we could pre-build images with the most used dependencies --similar to what @ericholscher said in our meeting.

Continuing with that line, now I'm mentioned that "only major version of node and minor version of rust are available", so we can generalize this dependency more in case we want to build those pre-built images.

I'm not sold on this, tho. However, I think in general that making users to specify dependencies is cleaner than asking them to run custom commands (*). Besides, it allow us to have better/standarized data and understand more how they are using our platform.

(*) if we ever change `nodenv` for a different `node` version manager, those commands will break.